_posts/not-all-changes-are-born-equal.md

```diff
@@ -0,0 +1,71 @@
+
+Let's step back for a second and resonate on coupling. First, zero coupling is impossible. Zero coupling is one of those utopic goals not that different from a completely secure computer. An utterly safe computer is not connected to anything, including a human being. A zero coupling architecture is an architecture that does nothing; there is no connection between all the parts composing it.
+
+When designing a system, the goal is to reduce coupling where it makes sense, accept coupling where it makes sense, encourage coupling where beneficial, and do not try to fight coupling where there is nothing to fight.
+
+### Hold on, can coupling be beneficial?
+
+Yes, it can. Let's use a logging library as a sample. If we're designing a component to be used by others, we need to allow users to plugin the logging library they want. In this situation, the traditional approach is to rely on an interface, something like `ILogger`. The component we're designing depends on `ILogger`, and the user, when using it, will provide an `ILogger` implementation. This approach decouples our code from the logging library implementation details, and users can provide their logging library, whose only requirement is to respect the interface contract.
+On the other hand, we want to couple code to the `ILogger` interface. If our code wasn't coupled to the interface, it means that we're not logging. Remember that zero coupling implies that we're doing nothing.
+
+### What kind of coupling do we want to avoid?
+
+The kind that forces us to make changes in portions of the codebase that, theoretically, should not be affected by the original change. When designing software, we encapsulate the data access logic so that any infrastructure change does not affect the infrastructure data access utilizer. The architecture uses encapsulation as a form of decoupling. The interface decouples the invoker's internal implementation; thus, any change to the implementation doesn't affect the caller. We can extend this sample to distributed systems and say that a shift in the internal behavior of a (micro)service should not affect any other (micro)service. If a cascade change occurs, they are coupled, which is probably not a good sign.
+
+A real-life sample would be something like this:
+
+> If a company stores its manufactures into a warehouse, whether the warehouse is automated or not is irrelevant for other departments, e.g., marketing.
+
+In other words, any change to the internal implementation of the warehouse is not visible from the outside. They are decoupled.
+Another sample is a new discount or pricing policy. If to change the pricing policy, there is a need to change multiple services, e.g., a products service and a sales service because product prices are in products, and the discount logic is in sales, it's clear that we have a coupling issue.
```

Comment:
```suggestion
Another example is a new discount or pricing policy. If to change the pricing policy, there is a need to change multiple services, e.g., a products service and a sales service because product prices are in products, and the discount logic is in sales, it's clear that we have a coupling issue.
```